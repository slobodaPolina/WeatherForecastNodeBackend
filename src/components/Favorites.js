import React from 'react';
import Forecast from './Forecast';
import SmallPreview from './SmallPreview';
import Loading from './Loading';
import { getCityByName, getTheFavorites, addCityToTheFavorites, deleteCityFromTheFavorites } from "../cityAdder";

export class Favorites extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            city: "", // it is the text from the searchBar
            isCityValid: true, // is the city user tried t add valid
            favorites: [] // array of objects of favorite cities. We keep only names in db
        };
        this.handleSearch = this.handleSearch.bind(this);
        this.handleInputChange = this.handleInputChange.bind(this);
        getTheFavorites(
            (data) => {
                data.forEach(name => {
                    getCityByName(
                        name,
                        (data) => {
                            this.addToTheFavorites(Object.assign(data, { loaded: true }));
                        },
                        () => {
                            console.error("Error getting favorite city " + name + " from weather API");
                            this.addToTheFavorites({
                                name,
                                loaded: false
                            });
                        }
                    );
                });
            },
            () => console.error("Problems getting list of favorite cities from db!")
        );
    }

    render() {
        let msg = this.state.isCityValid ? "" : <div className="error-message">Seems like the city is not valid</div>;
        return (
            <div>
                <div>
                    <div className="search-bar row">
                        <div className="bigText">Избранное</div>
                        <div>
                            <form onSubmit={this.handleSearch}>
                                <input id="search-input" type="search"
                                       placeholder="Добавить новый город" value={this.state.city} onChange={this.handleInputChange}/>
                                <button className="search-button roundButton">+</button>
                            </form>
                        </div>
                    </div>
                    {msg}
                </div>
                <div className="favorites">
                    {this.state.favorites.map((city) => {
                        return (
                            <div className="smallCityCard" key={city.name}>
                                <SmallPreview city={city} removeCity={this.removeGenerator(city.name)}/>
                                {city.loaded ? <Forecast data={city}/> : <Loading />}
                            </div>
                        )
                    })}
                </div>
            </div>
        );
    }

    handleInputChange(event) {
        this.setState({city: event.target.value});
    }

    handleSearch(event) {
        const getCallback = (isSuccess) => () => {
            this.setState({isCityValid: isSuccess});
            // todo remove loading card anyway
            if (isSuccess) { // means we got data from weather service
                addCityToTheFavorites(
                    this.state.city,
                    () => {
                        this.setState({favorites: this.state.favorites.push(this.state.city)});
                        this.setState({city: ""});
                    },
                    () => {
                        console.error('error during saving favorite city in db!');
                        this.setState({isCityValid: false});
                    }
                );
            }
        };

        event.preventDefault();
        if (this.props.favorites.find(city => city.name === this.state.city)) { // if we already have it we wont add it again
            this.setState({isCityValid: false});
        } else {
            // todo start the loader on the citycard
            // load real data and display it
            getCityByName(
                this.state.city,
                getCallback(true),
                getCallback(false)
            );
        }
    }

    removeGenerator(cityName) {
        return () => deleteCityFromTheFavorites(
            cityName,
            () => {
                this.setState({
                    favorites: this.state.favorites.filter(city => city.name !== cityName)
                })
            },
            () => { console.error("Problems removing city " + cityName + " from db!") }
         );
    }

    addToTheFavorites(obj) {
        this.setState({
            favorites: this.state.favorites.push(obj)
        });
    }
}

export default Favorites;
