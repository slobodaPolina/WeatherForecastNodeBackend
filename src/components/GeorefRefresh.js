import React from 'react';
import { connect } from 'react-redux';
import { addCityByGeolocation } from '../cityAdder';


export class GeorefRefresh extends React.Component {
    constructor(props) {
        super(props);
        this.addCity = this.addCity.bind(this);
    }

    addCity() {
        this.props.dispatch(addCityByGeolocation());
    }

    render() {
        return (
            <button className="georefRefresh"
                onClick={this.addCity}>
                    Обновить геолокацию
            </button>
        );
    }
}

export default connect()(GeorefRefresh);
