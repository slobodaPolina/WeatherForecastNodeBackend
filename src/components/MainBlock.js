import React from 'react';
import { connect } from 'react-redux';
import { addCityByGeolocation } from '../cityAdder';
import MainPreview from './MainPreview';
import Forecast from './Forecast';
import Loading from './Loading';

export class MainBlock extends React.Component {
    render() {
        if (this.props.data !== null) { // we know current geolocation
            let MINUTES_TO_BECOME_STALE = 30.0;
            let millisecDiff = Date.now() - (this.props.data.timeLoaded ? this.props.data.timeLoaded : new Date(0));
            let isOld = (millisecDiff / (1000 * 60)) > MINUTES_TO_BECOME_STALE;
            let oldCashedData = isOld ?
                <div className="error-message">
                    Please pay attention you see cashed stale data of your location. In order to update your geolocation or refresh cashed weather data, click the refresh button
                </div> : "";

            return (
                <div>
                    <div className="row main">
                        <MainPreview data={this.props.data}/>
                        <div className="width45">
                            <Forecast data={this.props.data}/>
                        </div>
                    </div>
                    {oldCashedData}
                </div>
            );
        } else {
            this.props.dispatch(addCityByGeolocation());
            return <Loading />;
        }
    }
}

export const MainBlockContainer = connect(
    state => ({ data: state.geolocation })
)(MainBlock);

export default MainBlockContainer;
