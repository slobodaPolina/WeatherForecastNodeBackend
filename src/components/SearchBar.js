import React from 'react';
import { addCityByName, getTheFavorites, addCityToTheFavorites } from '../cityAdder';

export class SearchBar extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            city: "",
            isCityValid: true,
            favorites: []
        };
        this.handleSearch = this.handleSearch.bind(this);
        this.handleInputChange = this.handleInputChange.bind(this);
        getTheFavorites(
            (data) => {this.state.favorites.push(data)},
            () => console.error("Problems getting favorite cities!")
        );
    }

    render() {
        let msg = this.state.isCityValid ? "" : <div className="error-message">Seems like the city is not valid</div>;
        return (
            <div>
                <div className="search-bar row">
                    <div className="bigText">Избранное</div>
                    <div>
                        <form onSubmit={this.handleSearch}>
                            <input id="search-input" type="search"
                                placeholder="Добавить новый город" value={this.state.city} onChange={this.handleInputChange}/>
                            <button className="search-button roundButton">+</button>
                        </form>
                    </div>
                </div>
                {msg}
            </div>
        );
    }

    handleInputChange(event) {
        this.setState({city: event.target.value});
    }

    handleSearch(event) {
        const getCallback = (isSuccess) => () => {
            this.setState({isCityValid: isSuccess});
            // remove loading card anyway
            this.props.dispatch({ //todo
                type: 'REMOVE_FAVORITE',
                cityName: this.state.city
            });
            if (isSuccess) { // means we got data from weather service
                addCityToTheFavorites(
                    this.state.city,
                    () => {
                        this.setState({favorites: this.state.favorites.push(this.state.city)});
                        this.setState({city: ""});
                    },
                    () => {
                        console.error('error during saving favorite city in db!');
                        this.setState({isCityValid: false});
                    }
                );
            }
        };

        event.preventDefault();
        if (this.props.favorites.find(cityName => cityName === this.state.city)) { // if we already have it we wont add it again
            this.setState({isCityValid: false});
        } else {
            // it is just to start the loader
            this.props.dispatch({ //todo deal with loading
                type: 'ADD_FAVORITE',
                city: {name: this.state.city}
            });
            // load real data and display it
            addCityByName(
                this.state.city,
                getCallback(true),
                getCallback(false)
            );
        }
    }
}

export default SearchBar;
