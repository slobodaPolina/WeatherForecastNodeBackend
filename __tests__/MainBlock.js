import React from 'react';
import renderer from 'react-test-renderer';
import { MainBlock } from '../src/components/MainBlock';

const data = {
    clouds: {
        all: 40
    },
    cod: 200,
    coord: {
        lat: 51.51,
        lon: -0.13
    },
    dt: 1573292996,
    id: 2643743,
    loaded: true,
    main: {
        humidity: 93,
        pressure: 1008,
        pressureAtm: "1.01",
        pressureMmHg: "756.00",
        temp: "2.48",
        temp_max: 277.59,
        temp_min: 274.26
    },
    name: "London",
    rain: {},
    sys: {
        country: "GB",
        id: 1414,
        sunrise: 1573283248,
        sunset: 1573316474,
        type: 1
    },
    timezone: 0,
    visibility: 5000,
    weather: [{
        description: "haze",
        icon: "50d",
        id: 721,
        main: "Haze"
    }],
    wind: {
        speed: 1.5
    }
}

it('renders mainBlock with no geolocation correctly', () => {
    const fn = jest.fn()
    const tree = renderer.create(
        <MainBlock data={null} dispatch={fn}/>
    ).toJSON();
    expect(tree).toMatchSnapshot();
});

it('renders MainBlock with geolocation correctly', () => {
  const tree = renderer.create(
          <MainBlock data={data}/>
  ).toJSON();
  expect(tree).toMatchSnapshot();
});
