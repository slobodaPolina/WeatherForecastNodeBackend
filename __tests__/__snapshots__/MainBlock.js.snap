// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders MainBlock with geolocation correctly 1`] = `
<div
  className="row main"
>
  <div
    className="width45 column"
  >
    <div
      className="mainTitle bigText"
    >
      London
    </div>
    <div
      className="row"
    >
      <img
        alt="Weather"
        className="weatherImage"
        src="/images/Haze.jpg"
      />
      <div
        className="hugeText"
      >
        2.48
        °C
      </div>
    </div>
  </div>
  <div
    className="width45"
  >
    <div
      className="forecastContainer column"
    >
      <div
        className="forecastRow row"
      >
        <div
          className="paramTitle"
        >
          Ветер
        </div>
        <div
          className="paramValue"
        >
          1.5
           meter/sec
        </div>
      </div>
      <div
        className="forecastRow row"
      >
        <div
          className="paramTitle"
        >
          Облачность
        </div>
        <div
          className="paramValue"
        >
          haze
        </div>
      </div>
      <div
        className="forecastRow row"
      >
        <div
          className="paramTitle"
        >
          Давление
        </div>
        <div
          className="paramValue"
        >
          756.00
           mm Hg
        </div>
      </div>
      <div
        className="forecastRow row"
      >
        <div
          className="paramTitle"
        >
          Влажность
        </div>
        <div
          className="paramValue"
        >
          93
          %
        </div>
      </div>
      <div
        className="forecastRow row"
      >
        <div
          className="paramTitle"
        >
          Координаты
        </div>
        <div
          className="paramValue"
        >
          [
          -0.13
          , 
          51.51
          ]
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`renders mainBlock with no geolocation correctly 1`] = `
<div>
  <div
    className="bigText"
  >
    Подождите, данные загружаются
  </div>
  <div
    className="refreshBanner"
  >
    <img
      alt="loading"
      className="refreshImage"
      src="/images/refresh.svg"
    />
  </div>
</div>
`;
